lib/ddtrace/contrib/httprb/datadog_wrap.rb

```diff
@@ -0,0 +1,109 @@
+
+require 'uri'
+require 'ddtrace/ext/app_types'
+require 'ddtrace/ext/http'
+require 'ddtrace/ext/net'
+require 'ddtrace/ext/distributed'
+require 'ddtrace/contrib/analytics'
+require 'ddtrace/propagation/http_propagator'
+
+require 'http'
+
+module Datadog
+  module Contrib
+    module Httprb
+      class DatadogWrap < ::HTTP::Feature
+
+        def initialize(*)
+          super
+        end
+
+        def wrap_request(request)
+          begin
+            return super(request) unless tracer_enabled?
+
+            datadog_span = datadog_configuration[:tracer].trace(
+              Ext::SPAN_REQUEST,
+              service: datadog_configuration[:service_name],
+              span_type: Datadog::Ext::HTTP::TYPE_OUTBOUND
+            )
+            
+            if request && request.headers
+              Datadog::HTTPPropagator.inject!(datadog_span.context, request.headers)
+            end
+
+            Contrib::Analytics.set_sample_rate(datadog_span, analytics_sample_rate) if analytics_enabled?
+
+            http_method = request.try(:verb).try(:to_s).try(:upcase)
+
+            datadog_span.resource = http_method  
+
+            if request.try(:uri)
+              uri = request.uri
+              datadog_span.set_tag(Datadog::Ext::HTTP::URL, uri.path)
+              datadog_span.set_tag(Datadog::Ext::HTTP::METHOD, http_method)
+              datadog_span.set_tag(Datadog::Ext::NET::TARGET_HOST, uri.host)
+              datadog_span.set_tag(Datadog::Ext::NET::TARGET_PORT, uri.port)
+            end
+
+            return request
+          ensure
+            return request
+          end
+        end
+
+        def wrap_response(response)
+          begin
+            return super(request) unless tracer_enabled?
+
+            tracer = datadog_configuration[:tracer]
+            datadog_span = tracer.active_span
```

Comment:
Question for @delner: what's our way to ensuring that `tracer.active_span` is the trace we expect in integrations like these (with separate start and end events)?

It seems like it would find the correct one no child integration "leaks", but I'm not sure how confident we are in this.

Reply:
I think this is a good observation @marcotc; I don't think this is a safe operation because although in theory, if everything else is working as intended it should be no problem. If a span is opened (but not closed) between `wrap_request` and `wrap_response`, this would return the wrong span.

If there's a better way to 'track' the span without leaking, we should do that instead. Using `tracer.trace` with a block is the most reliable way. I think Marco suggested the use of ActiveSupport notifications elsewhere, for which we already have a module that serves as a reliable means of doing this work for you: so we might want to consider that.

Reply:
per latest comments, refactored to monkeypatching + wrapping with `.trace`, rather than AS approach